src/async/mod.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: remove following line once all changes are merged into master
+#![allow(dead_code)]
+
+mod promise;
+mod lock;
+
+use std::fmt::{self, Debug, Formatter};
+use std::sync::Arc;
+
+use futures::{Async, Future, Poll};
+use futures::task::{self, Task};
+
+use error::{Error, Result};
+use self::lock::SpinLock;
+use self::promise::Shutdown as ShutdownPromise;
+
+
+/// A handle that is used to notify future that the task finishes.
+pub struct NotifyHandle<T> {
```

Comment:
No, it's not. It's a reference to the underlying result and task. `Promise` is more like a handler, which invokes specific callback.